packages/mongodb-dialects/spring-criteria/src/test/kotlin/com/mongodb/jbplugin/dialects/springcriteria/IntegrationTest.kt

```diff
@@ -0,0 +1,210 @@
+                .createCodeInsightFixture(
+                    projectFixture,
+                )
+
+        context.getStore(namespace).put(testFixtureKey, testFixture)
+        testFixture.setUp()
+
+        ApplicationManager.getApplication().invokeAndWait {
+            val module = testFixture.module
+
+            if (!JavaLibraryUtil.hasLibraryJar(module, "org.mongodb:mongodb-driver-sync:5.1.0")) {
+                runCatching {
+                    PsiTestUtil.addProjectLibrary(
+                        module,
+                        "org.springframework.data:spring-data-mongodb:4.3.2",
+                        listOf(pathToClassJarFile(MongoTemplate::class.java)),
+                    )
+
+                    PsiTestUtil.addProjectLibrary(
+                        module,
+                        "org.springframework.data:spring-data-mongodb-mapping:4.3.2",
+                        listOf(pathToClassJarFile(Document::class.java)),
+                    )
+                }
+            }
+        }
+
+        PsiTestUtil.addSourceRoot(testFixture.module, testFixture.project.guessProjectDir()!!)
+        val tmpRootDir = testFixture.tempDirFixture.getFile(".")!!
+        PsiTestUtil.addSourceRoot(testFixture.module, tmpRootDir)
+        context.getStore(namespace).put(testPathKey, tmpRootDir.path)
+    }
+
+    override fun beforeEach(context: ExtensionContext) {
+        val fixture = context.getStore(namespace).get(testFixtureKey) as CodeInsightTestFixture
+        val modulePath = context.getStore(namespace).get(testPathKey).toString()
+
+        ApplicationManager.getApplication().invokeAndWait {
+            val parsingTest = context.requiredTestMethod.getAnnotation(ParsingTest::class.java) ?: return@invokeAndWait
+
+            val fileName = Path(modulePath, "src", "main", "java", parsingTest.fileName).absolutePathString()
+
+            fixture.configureByText(
+                fileName,
+                parsingTest.value,
+            )
+        }
+    }
+
+    override fun interceptTestMethod(
+        invocation: InvocationInterceptor.Invocation<Void>,
+        invocationContext: ReflectiveInvocationContext<Method>,
+        extensionContext: ExtensionContext,
+    ) {
+        val throwable: AtomicReference<Throwable?> = AtomicReference(null)
+        val finished = AtomicBoolean(false)
+
+        val fixture = extensionContext.getStore(namespace).get(testFixtureKey) as CodeInsightTestFixture
+        val dumbService = DumbService.getInstance(fixture.project)
+
+        dumbService.runWhenSmart {
+            val result =
+                runCatching {
+                    invocation.proceed()
+                }
+            result.onSuccess {
+                finished.set(true)
+            }
+
+            result.onFailure {
+                finished.set(true)
+                throwable.set(it)
+            }
+        }
+
+        while (!finished.get()) {
+            Thread.sleep(1)
+        }
```

Comment:
I guess there's no good alternative to busy-looping here?

Reply:
Sadly we need to wait for the other thread to finish. There are other ways (maybe we can use a runBlocking coroutine). Let me take a look, it might be easier to understand and maintain.

Reply:
@addaleax Changed it to use a coroutine, WDYT?

Reply:
yep, that looks better! :)